.copilot/general_instructions.md

```diff
@@ -0,0 +1,119 @@
+* **Target Audience:** The code is maintained by a large, distributed team of professional developers. Code quality is paramount to ensure the project's long-term health.
+
+# 2. General Code Philosophy
+
+* **Performance is a Feature:** The core code is written in C++ with an explicit focus on performance. Code should be efficient, avoid unnecessary allocations and copies, and leverage the hardware effectively.
+* **Correctness over All Else:** A database must be correct. Suggestions that prioritize performance at the cost of correctness (e.g., race conditions, data corruption) must be avoided.
+* **Readability is Key:** Code should be easy to read and understand. This is essential for a project with many contributors. Avoid overly clever or obscure constructs.
+* **Self-Explanatory Code:** The code should be its own documentation. Copilot should aim to generate code where the intent is clear from the variable and function names. Comments should be used sparingly, only to explain *why* something is done, not *what* is being done.
+
+# 3. Naming Conventions
+
+* **Be Descriptive:** Names must be as self-explanatory as possible.
+    * **Bad:** `int x;` (What is `x`?)
+    * **Good:** `int row_count;`
+* **Full Words Preferred:** Prefer full, descriptive words over abbreviations.
+    * **Bad:** `std::string prc_tbl;`
+    * **Good:** `std::string processed_table;`
+* **Consistent Casing:** Use a consistent casing style. C++ code should follow the project's established style (`snake_case` for variables). Python test code should follow PEP 8 (`snake_case`).
+* **Unit and Purpose:** Naming should reflect the unit of measurement or the purpose of the variable.
+    * **Bad:** `auto timeout = 5;` (Is this 5 seconds? Milliseconds? What is it for?)
+    * **Good:** `std::chrono::seconds request_timeout = 5s;`
+
+# 4. Code Structure and Style
+
+* **Follow Established Style:** When working on existing files, adhere to the established style of that file. This overrides any personal preference.
+* **Modern C++ and Python:** Favor modern language features that improve clarity and safety.
+    * **C++:** C++17 or C++20 is the standard. Prefer `std::optional`, `std::variant`, and smart pointers (`std::unique_ptr`, `std::shared_ptr`) for resource management.
+    * **Python:** Follow PEP 8 and use modern Python features (e.g., f-strings, type hints).
+* **Avoid Code Duplication:** Copilot should strive to avoid generating redundant code. If a common pattern exists, it should suggest calling or refactoring to a shared utility function.
+
+# 5. Documentation and Comments
+
+* **Docstrings:** Public APIs (classes, functions) in C++ headers and Python modules should be documented with a brief explanation of their purpose, parameters, and return values.
+* **In-Code Comments:** Comments should be used as a last resort. If a comment is needed to explain a piece of code, it may be a sign that the code is too complex and should be simplified or refactored.
```

Comment:
I don't agree with this statement, but I know some other people do.
This line also contradicts the next one - the next line says it's good to write a comment to "explain a non-obvious algorithm", but this line says that it's "a sign that the code is too complex and should be simplified or refactored". An algorithm can be non-obvious without needing to be simplified or refactored.

Reply:
Agree with @nyh. Sometimes comments can be abused to explain convoluted code which should otherwise be refactored. But sometimes they explain a complex situation or algorithm.